### PR TITLE
Add more Jackson dependencies to spring-boot-dependencies

### PR DESCRIPTION
**Pull request by candrews**
Add jackson-module-parameter-names, jackson-datatype-hibernate3, jackson-datatype-hibernate4, jackson-datatype-hibernate5, and jackson-datatype-jdk7.

I have signed and agree to the terms of the SpringSource Individual Contributor License Agreement.

### CODE REVIEW

File: spring-boot-dependencies/pom.xml [hunk 1]
```diff
@@ -535,6 +535,31 @@
 				<version>${jackson.version}</version>
 			</dependency>
 			<dependency>
+				<groupId>com.fasterxml.jackson.datatype</groupId>
+				<artifactId>jackson-datatype-jdk7</artifactId>
+				<version>${jackson.version}</version>
+			</dependency>
+			<dependency>
+				<groupId>com.fasterxml.jackson.datatype</groupId>
+				<artifactId>jackson-datatype-hibernate3</artifactId>
```
Comment (wilkinsona) [inline spring-boot-dependencies/pom.xml]: Given that Boot uses Hibernate 4, I'm not sure that it makes sense to add this

File: spring-boot-dependencies/pom.xml [hunk 2]
```diff
@@ -535,6 +535,31 @@
...
+				<version>${jackson.version}</version>
+			</dependency>
+			<dependency>
+				<groupId>com.fasterxml.jackson.datatype</groupId>
+				<artifactId>jackson-datatype-hibernate5</artifactId>
+				<version>${jackson.version}</version>
+			</dependency>
+			<dependency>
+				<groupId>com.fasterxml.jackson.module</groupId>
+				<artifactId>jackson-module-parameter-names</artifactId>
```
Comment (wilkinsona) [inline spring-boot-dependencies/pom.xml]: See also #3615. The main [Jackson README](https://github.com/fasterxml/jackson#other-modules-experimental) lists the module as experimental, although the module's author says it should be stable now. I'm still a little wary of adding dependency management for it, though.

**Comment by candrews** [comment]
@wilkinsona I can see your point about hibernate3 module - I'll remove it.

However, for the others, we're just basically allowing users to use them without manually specifying their versions so that the Jackson version remains consistent. We're not actually pulling in the dependencies or otherwise using (or even encouraging) their use - just making it easier to use if the user so desires. Therefore, I'm a bit unclear about what the concern is with adding them to the BOM.

**Comment by wilkinsona** [comment]
> I'm a bit unclear about what the concern is with adding them to the BOM.

We take backwards compatibility seriously. Every new dependency in the bom is one whose versions we need to track and that we need to keep up-to-date. Generally speaking, Jackson's backwards compatibility is excellent. However, I don't know if they follow different rules for experimental modules hence me being wary of `jackson-module-parameter-names`.

**Comment by lpandzic** [comment]
Hello, jackson-module-parameter-names author here.

I don't expect many changes to the module in the future since the core functionality was done over a year ago when Java 8 came out. Code changes that happened later resolved around retrofitting the module to the changes in the Jackson API for introspectors and fixing/investigating several bugs.
All changes to the module are approved and merged by Jackson lead.
I hope that gives you some reassurance about backwards compatibility.

In fact, the module upholds the backwards compatibility for delegating creators, which [annoyed some users](https://github.com/FasterXML/jackson-module-parameter-names/issues/21). I added an option to [override this behavior](https://github.com/FasterXML/jackson-module-parameter-names#delegating-creator).

**Comment by wilkinsona** [comment]
Thanks, @lpandzic. I'm convinced.

@candrews I'll take care of removing `jackson-datatype-hibernate3` as part of merging this; no need for you to do anything.

**Comment by wilkinsona** [comment]
Merged into master. Thanks for the PR, @candrews.

**Pull request closed without merging**